Tests/ApolloCodegenTests/CodeGenIR/IRRootEntityFieldBuilderTests.swift

```diff
@@ -2937,6 +3118,396 @@ class IRRootEntityFieldBuilderTests: XCTestCase {
       .to(shallowlyMatch(allAnimals_asWarmBlooded_height_expected))
   }
 
+  func test__mergedSelections__givenSiblingTypeCasesAndNestedEntityTypeCases_onlyNestedEntityFieldMergeTypeCases() throws {
+    // given
+    schemaSDL = """
+    type Query {
+      allAnimals: [Animal!]
+    }
+
+    interface Animal {
+      species: String
+    }
+
+    interface Pet {
+      predator: Animal
+      height: Height
+    }
+
+    interface Reptile {
+      skinCovering: String
+    }
+
+    type Cat implements Pet & Animal {
+      species: String
+      breed: String
+      height: Height
+      predator: Animal
+    }
+
+    type Height {
+      feet: Int
+    }
+    """
+
+    document = """
+    query Test {
+      allAnimals {
+        ... on Pet {
+          ... on Reptile {
+            skinCovering
+          }
+          predator {
+            ... on Pet {
+              height {
+                feet
+              }
+            }
+            ... on Reptile {
+              skinCovering
+            }
+            ... on Cat {
+              breed
+            }
+          }
+        }
+        ... on Reptile {
+          skinCovering
+        }
+        ... on Cat {
+          breed
+        }
+      }
+    }
+    """
+    
+    // when
+    try buildSubjectRootField()
+
+    let allAnimals = subject[field: "allAnimals"]
+
+    expect(allAnimals?[as: "Cat"]?[field: "predator"]?[as: "Pet"]).toNot(beNil())
+    expect(allAnimals?[as: "Cat"]?[field: "skinCovering"]).to(beNil())
+    expect(allAnimals?[as: "Cat"]?[as: "Reptile"]).to(beNil())
+    expect(allAnimals?[as: "Cat"]?[as: "Pet"]).to(beNil())
+
+    expect(allAnimals?[as: "Cat"]?[field: "predator"]?[as: "Pet"]?[field: "height"]).toNot(beNil())
+    expect(allAnimals?[as: "Cat"]?[field: "predator"]?[as: "Pet"]?[field: "skinCovering"]).to(beNil())
+    expect(allAnimals?[as: "Cat"]?[field: "predator"]?[as: "Pet"]?[field: "breed"]).to(beNil())
```

Comment:
`predator.asCat` should have a height field, because Cat implements Pet. But `predator.asPet` should not have a `breed` field, because `Pet` does not implement `Cat`.